### PR TITLE
Added FrozenLakeEnv

### PR DESCRIPTION
**Pull request by jaraujo98**
Added the Frozen Lake Environment.

### CODE REVIEW

File: src/backend/FrozenLakeEnv.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    this.nA = 4;
+    this.nS = this.nrow * this.ncol;
+
+    this.reset();
+  }
+
+	// Returns a tensor describing the state space of the env
+	// Will be used to determine the input dimensionality of the model
+	getStateSpace() {
+    return 2;
```
Comment (eito-fis) [inline src/backend/FrozenLakeEnv.js:52]: Shouldn't this return `this.nS`?

File: src/backend/FrozenLakeEnv.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+	// Returns a tesor describing the action space of the env
+	// Will be used to determine the output dimensionality of the model
+	getActionSpace() {
+    return 4;
```
Comment (eito-fis) [inline src/backend/FrozenLakeEnv.js:58]: Same as above, should return `this.nA`?

File: src/backend/FrozenLakeEnv.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+	// Returns the current state
+	getCurrentState() {
+    return this.state;
+  }
+
+	// Takes an action as an integer and updates the state
+	step(action) {
+    if (![1,2,3,4].includes(action)) {
```
Comment (eito-fis) [inline src/backend/FrozenLakeEnv.js:74]: 0 should be a valid action

File: src/backend/FrozenLakeEnv.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    } else if (action === 3) { // Up
+      row = Math.max(row-1, 0);
+    } else {
+      alert("Invalid action!");
+    }
+
+    this.state = [row, col];
+
+    if (this.desc[row][col] === "H") {
+      env.reset();
```
Comment (eito-fis) [inline src/backend/FrozenLakeEnv.js:96]: I think we want to set a done flag here, and leave the resetting to the `rollout` function

**Review by eito-fis — COMMENTED** [review]
Looking good, some changes I think should be made.

File: src/backend/FrozenLakeEnv.js (hunk 1 above)
Comment (jaraujo98) [reply src/backend/FrozenLakeEnv.js:52]: That's a good point. The state is a vector [x, y], so I just returned its shape (should have been [1, 2], now that I think of it....). But I was really sleepy, that doesn't make sense. Maybe returning [4,4] (or [8,8]) would make more sense?

File: src/backend/FrozenLakeEnv.js (hunk 2 above)
Comment (jaraujo98) [reply src/backend/FrozenLakeEnv.js:58]: Maybe return a vector with the possible actions? I know we discussed this yesterday. If I recall correctly, you only wanted the number of actions. In that case, `this.nA` would do the work.

File: src/backend/FrozenLakeEnv.js (hunk 3 above)
Comment (jaraujo98) [reply src/backend/FrozenLakeEnv.js:74]: Nice catch!

File: src/backend/FrozenLakeEnv.js (hunk 1 above)
Comment (eito-fis) [reply src/backend/FrozenLakeEnv.js:52]: It should return whatever the shape of the input to the model will be. If we're just gonna feed the 2D coords into the model then the shape can be 2 or [2] I think

File: src/backend/FrozenLakeEnv.js (hunk 2 above)
Comment (eito-fis) [reply src/backend/FrozenLakeEnv.js:58]: Yeah, same as above I really just want something to stick into the models. Integer number of actions would be easiest

File: src/backend/FrozenLakeEnv.js (hunk 4 above)
Comment (jaraujo98) [reply src/backend/FrozenLakeEnv.js:96]: Makes sense

**Pull request merged**